migration/idempotent/018-metric-rollup-maintenance.sql

```diff
@@ -0,0 +1,155 @@
+    r RECORD;
+
+BEGIN
+    SET LOCAL search_path = pg_catalog, pg_temp;
+    _refresh_interval := config ->> 'refresh_interval';
+    FOR r IN
+        SELECT table_schema, table_name, rollup_id FROM _prom_catalog.metric
+           WHERE is_view AND view_refresh_interval = _refresh_interval
+    LOOP
+        IF (r.rollup_id IS NOT NULL AND NOT _refresh_rollups) THEN
+            -- Do not refresh the rollups if the user does not want automatic downsampling.
+            CONTINUE;
+        END IF;
+        CALL public.refresh_continuous_aggregate(r.table_schema || '.' || r.table_name, current_timestamp - 3 * _refresh_interval, current_timestamp);
+        COMMIT; -- Commit after every refresh to avoid high I/O & mem-buffering.
+        SET LOCAL search_path = pg_catalog, pg_temp;
+    END LOOP;
+END;
+$$
+LANGUAGE PLPGSQL;
+COMMENT ON PROCEDURE _prom_catalog.caggs_refresher IS 'caggs_refresher runs every refresh_interval passed in config. Its
+main aim is to refresh those Caggs that have been registered under _prom_catalog.metric and whose view_refresh_interval
+matches the given refresh_interval. It refreshes 2 kinds of Caggs:
+1. Caggs created by metric rollups
+2. Custom Caggs created by the user';
+
+CREATE OR REPLACE PROCEDURE _prom_catalog.create_cagg_refresh_job_if_not_exists(_refresh_interval INTERVAL)
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+BEGIN
+    IF (
+        SELECT EXISTS (
+            SELECT 1 FROM timescaledb_information.jobs WHERE
+               proc_name = 'caggs_refresher' AND schedule_interval = _refresh_interval
+        )
+    ) THEN
+        -- Refresh job exists for the given _refresh_interval. Hence, nothing to do.
+        RETURN;
+    END IF;
+    PERFORM public.add_job('_prom_catalog.caggs_refresher', _refresh_interval, FORMAT('{"refresh_interval": "%s"}', _refresh_interval)::jsonb);
+END;
+$$
+LANGUAGE PLPGSQL;
+COMMENT ON PROCEDURE _prom_catalog.create_cagg_refresh_job_if_not_exists(INTERVAL) IS
+'Creates a Cagg refresh job that refreshes all Caggs registered by register_metric_view().
+This function creates a refresh job only if no caggs_refresher() exists currently with the given refresh_interval.';
+
+CREATE OR REPLACE PROCEDURE _prom_catalog.caggs_compressor(job_id int, config jsonb)
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+DECLARE
+    r RECORD;
+    _temp INTEGER;
+
+BEGIN
+    FOR r IN
+        SELECT
+            m.table_schema || '.' || m.table_name AS compressible_cagg,
+            m.view_refresh_interval AS refresh_interval
+        FROM
+            _prom_catalog.metric m
+        INNER JOIN
+            timescaledb_information.continuous_aggregates cagg
+        ON (m.table_name = cagg.view_name AND m.table_schema = cagg.view_schema)
+        WHERE
+            cagg.compression_enabled AND m.is_view
+    LOOP
+        SELECT count(*) INTO _temp FROM (SELECT public.compress_chunk(public.show_chunks(r.compressible_cagg, older_than => 100 * r.refresh_interval))) a;
```

Comment:
> why count into _temp?

There used to be a log saying how many chunks compressed/dropped, but then I removed it so that sql-tests snapshots match (Noticed that PG 13 and PG 14 have different chunks, hence had to remove).


> why is it 100 * refresh_interval ?

Since we refresh every refresh_interval, we essentially have 1 sample every refresh_interval. Doing `100 * refresh_interval` ensures that we compress only when we have atleast a 100 samples. Otherwise, we don't compress.

> Do we not have a way to set retention policies on caggs?

There isn't anything related to retention here. That said, we handle retention manually, by dropping chunks, than depending on retention policy per cagg. This is for perf reasons.